ARROW-13643: [C++][Compute] Implement outer join with support for residual predicates

save-buffer opened the pull request:
Implements residual predicates on hash join

save-buffer commented:
The CI failure seems to be a failure with failing to download python 3.1 (https://discuss.python.org/t/disappearing-macos-packages-on-python-org/11737), which is unrelated to this PR.

bkietz commented on cpp/src/arrow/compute/exec/hash_join.h:
```diff
@@ -56,19 +56,36 @@ class ARROW_EXPORT HashJoinSchema {
   std::shared_ptr<Schema> MakeOutputSchema(const std::string& left_field_name_prefix,
                                            const std::string& right_field_name_prefix);
 
+  bool HasLeftPayload() { return HasPayload(0); }
```
Nit: could you replace this with `LeftPayloadIsEmpty` for greater specificity?

save-buffer replied:
Actually thinking about this one more too, I think it would be better to leave it as some sort of positive, as the code paths that use this all check if a payload exists, not if a payload doesn't exist. A different name could be "LeftPayloadExists" or "LeftPayloadIsNonempty".

bkietz reviewed (changes requested):
Thanks for doing this! I'm still reading this PR, but I've got a few minor comments now

save-buffer commented on cpp/src/arrow/compute/exec/hash_join_node.cc:
```diff
@@ -255,17 +284,83 @@ std::shared_ptr<Schema> HashJoinSchema::MakeOutputSchema(
+
+Result<std::vector<FieldRef>> HashJoinSchema::CollectFilterColumns(
+    const Expression& filter, const Schema& schema) {
+  std::vector<FieldRef> nonunique_refs;
+  RETURN_NOT_OK(TraverseExpression(nonunique_refs, filter, schema));
+
+  std::vector<FieldRef> result;
+  std::unordered_set<int> seen_paths;
+  for (auto ref : nonunique_refs) {
+    ARROW_ASSIGN_OR_RAISE(auto match, ref.FindOne(schema));
+    if (seen_paths.find(match[0]) == seen_paths.end()) {
+      seen_paths.insert(match[0]);
+      result.push_back(ref);
+    }
+  }
+  return result;
```
The problem is it's unclear what an index FieldRef would refer to in this situation. In particular, you can't distinguish between FieldRefs pointing to build-side columns and those pointing to probe-side columns.

bkietz replied:
Isn't it the case that the output schema of a hash join node is always
```python
schema({
  **left_payload,
  **left_keys,
  **right_payload,
  **right_keys,
})
```
In light of this, I think it's possible to map predictably from an index in that schema to a field in one of the inputs.

Addendum to one of the unit tests making this explicit:
```diff
diff --git a/cpp/src/arrow/compute/exec/plan_test.cc b/cpp/src/arrow/compute/exec/plan_test.cc
index 437a93f9e..c250b305d 100644
--- a/cpp/src/arrow/compute/exec/plan_test.cc
+++ b/cpp/src/arrow/compute/exec/plan_test.cc
@@ -1046,6 +1046,13 @@ TEST(ExecPlanExecution, SelfInnerHashJoinSink) {
         auto hashjoin,
         MakeExecNode("hashjoin", plan.get(), {left_filter, right_filter}, join_opts));

+    ASSERT_EQ(*hashjoin->output_schema(), Schema({
+                                              field("l_i32", int32()),
+                                              field("l_str", utf8()),
+                                              field("r_i32", int32()),
+                                              field("r_str", utf8()),
+                                          }));
+
     ASSERT_OK_AND_ASSIGN(std::ignore, MakeExecNode("sink", plan.get(), {hashjoin},
                                                    SinkNodeOptions{&sink_gen}));

```

save-buffer replied:
The problem is if a filter column is not an output column. This scheme would work if we used indices into "concatenated input schema" (see the other comment by Weston about this)

bkietz replied:
Ah, I see the problem. Concatenated input schema would work, alternatively
```python
schema({
  **left_payload,
  **left_keys,
  **left_filtered,
  **right_payload,
  **right_keys,
  **right_filtered,
})
```
where `_filtered` fields are referenced by the filter expression but are neither keys nor part of the payload

bkietz replied:
Nvm `_filtered`; concatenated is what substrait uses so it'll be most straightforward to stick with that.

save-buffer replied:
That sounds good to me.

save-buffer commented on cpp/src/arrow/compute/exec/expression.h:
```diff
@@ -91,6 +91,9 @@ class ARROW_EXPORT Expression {
   bool IsSatisfiable() const;
 
+  /// Return true if this expression has no clauses.
+  bool IsEmpty() const;
```
Thinking about it more, I'm not a big fan of this `literal(true)` convention. Currently, null expressions are allowed (by calling the default constructor). It's very bug-prone leaving code around that assumes expressions are non-null while having no good way of checking this. Maybe instead of introducing `IsEmpty` you can do `== Expression()`, but at that point I feel `IsEmpty` is cleaner. 

But anyway, If this is the convention, then the default constructor should be deleted to eliminate the possibility of there ever being a null `Expression`.

bkietz replied:
`Expression()` is currently allowed to enable preallocation in `std::vector<Expression>`. IIUC, the only circumstances under which an expression would be empty are (a) a preallocated element of a container which will be overwritten shortly (b) a moved-from variable, *any* use of which should be avoided or (c) when used as a special-case placeholder for non-empty expressions.

If I'm missing something, please file a follow up JIRA for adding `IsEmpty()`.

save-buffer replied:
Well I think it's more of just a gotcha kind of thing. You're allowed to create an object that's in an invalid state, and the rest of the codebase assumes it's not in that state. My thinking is that if you're allowed to create an object in that state, it should be considered a valid value, and should therefore be checked against.

bkietz replied:
If it were possible to forbid default construction of `Expression` without breaking `vector<Expression>::resize` I would do so.  As I've been trying to express, it's not really valid for humans to write `Expression defaulted;` but there's no attribute or pragma I'm aware of to make the compiler warn about that. Maybe the solution is to add a large warning comment for `Expression() = default;`

save-buffer replied:
Right, so I'd argue the solution is to make `Expression defaulted;` valid. Anyway, this is probably out of scope for this PR.

westonpace commented on cpp/src/arrow/compute/exec/hash_join.cc:
```diff
@@ -282,6 +279,125 @@ class HashJoinBasicImpl : public HashJoinImpl {
+      bool passed = false;
+      for (; static_cast<size_t>(irow) < num_rows && match_left[irow] == curr_left;
+           irow++) {
+        bool is_null = !BitUtil::GetBit(nulls, irow);
+        bool is_cmp_true = BitUtil::GetBit(comparisons, irow);
+        // We treat a null comparison result as false, like in SQL
+        if (!is_null && is_cmp_true) {
```
Super minor nit:  `is_valid` might be clearer to read since you wouldn't have a double negation.

westonpace commented on cpp/src/arrow/compute/exec/hash_join_node.cc:
```diff
@@ -43,32 +43,49 @@ bool HashJoinSchema::IsTypeSupported(const DataType& type) {
+    payload_fields.erase(match[0]);
   }
 
+  std::vector<FieldRef> result;
```
Minor nit: I know it was this way before but in general it's nice if we can avoid names like `result` and use something meaningful like `payload_refs`

westonpace commented on cpp/src/arrow/compute/exec/hash_join_node.cc:
```diff
@@ -43,32 +43,49 @@ bool HashJoinSchema::IsTypeSupported(const DataType& type) {
-  for (size_t i = 0; i < b.size(); ++i) {
-    ARROW_ASSIGN_OR_RAISE(auto match, b[i].FindOne(schema));
-    b_paths.insert(match[0]);
+Result<std::vector<FieldRef>> HashJoinSchema::ComputePayload(
+    const Schema& schema, const std::vector<FieldRef>& output,
+    const std::vector<FieldRef>& filter, const std::vector<FieldRef>& keys) {
+  // payload = (output + filter) - keys, with no duplicates
```
So what happens if there are duplicate field names in the schema?  Is this a malformed query?

@bkietz Is this something the IR would detect and prevent in the first place?

bkietz replied:
IR requires that fields be referenced by index, so it would be up to a producer to decide how to navigate duplicate field names.

westonpace replied:
Ok.  I spoke to Sasha a bit about this as well.  Just to close the loop, in IR, in a join expression, field reference indices are indices into the concatenated {left_schema, right_schema} correct?  So `field_ref(5)` is `left_schema[5]` if `len(left_schema) > 5` and otherwise it is `right_schema[5 % len(left_schema)]`?

save-buffer replied:
Should be `right_schema[5 - len(left_schema)]` (the pathological case is if `len(left_schema) == 1`, then using `%` would always evaluate to `0`).

bkietz replied:
Concatenated schema as you describe is what substrait specifies https://substrait.io/relations/logical_relations/#join-operation so we should probably stick with that for simplicity.

westonpace commented on cpp/src/arrow/compute/exec/hash_join_node.cc:
```diff
@@ -274,17 +303,83 @@ std::shared_ptr<Schema> HashJoinSchema::MakeOutputSchema(
+Result<Expression> HashJoinSchema::BindFilter(Expression filter,
+                                              const Schema& left_schema,
+                                              const Schema& right_schema) {
+  if (filter.IsBound()) {
+    return std::move(filter);
+  }
```
Maybe `if (filter.IsBound() || filter == literal(true))` and then you could get rid of the next if.

westonpace commented on cpp/src/arrow/compute/exec/hash_join_node.cc:
```diff
@@ -274,17 +303,83 @@ std::shared_ptr<Schema> HashJoinSchema::MakeOutputSchema(
+  return result;
+}
+
+Status HashJoinSchema::TraverseExpression(std::vector<FieldRef>& refs,
+                                          const Expression& filter,
+                                          const Schema& schema) {
+  if (filter == literal(true)) return Status::OK();
+  if (auto* call = filter.call()) {
+    for (const Expression& arg : call->arguments)
+      RETURN_NOT_OK(TraverseExpression(refs, arg, schema));
+  } else if (auto* param = filter.parameter()) {
+    if (!param->ref.IsName())
+      return Status::Invalid("Filter parameters to join must be by name");
+    ARROW_ASSIGN_OR_RAISE(auto match, param->ref.FindOneOrNone(schema));
+    if (match != FieldPath()) refs.push_back(param->ref);
+  }
+  return Status::OK();
+}
+
```
Seems like this could be some kind of utility method as part of `Expression`.

westonpace commented on cpp/src/arrow/compute/exec/hash_join_node.cc:
```diff
@@ -300,30 +395,36 @@ class HashJoinNode : public ExecNode {
           join_options.output_prefix_for_left, join_options.output_prefix_for_right));
     }
 
+    ARROW_ASSIGN_OR_RAISE(Expression filter,
+                          schema_mgr->BindFilter(std::move(join_options.filter),
+                                                 left_schema, right_schema));
+
```
I think you'd get away with it 9 times out of 10 but I'm not sure it is safe to move `join_options.filter`

westonpace commented on cpp/src/arrow/compute/exec/hash_join_node_test.cc:
```diff
@@ -1061,17 +1066,17 @@ TEST(HashJoin, Random) {
 
     // Print test case parameters
     // print num_rows, batch_size, join_type, join_cmp
-    std::cout << join_type_name << " " << key_cmp_str << " ";
+    std::cout << "Trial " << test_id << ":\n";
+    std::cout << "    " << join_type_name << " " << key_cmp_str << " ";
```
I'm a little surprised we have print statements here.  I think we normally avoid this kind of thing for tests, favoring `ARROW_SCOPED_TRACE` instead but since they were here already :shrug:

bkietz replied:
It isn't the preferred style, but it's not necessary to address this here. I've added
https://issues.apache.org/jira/browse/ARROW-14641

westonpace commented on cpp/src/arrow/compute/exec/hash_join.cc:
```diff
@@ -282,6 +279,125 @@ class HashJoinBasicImpl : public HashJoinImpl {
+
+    size_t num_probed_rows = match.size() + no_match.size();
+    if (mask.is_scalar()) {
+      const auto& mask_scalar = mask.scalar_as<BooleanScalar>();
```
Is this branch covered by your testing?

save-buffer replied:
I've added a test to cover it (`TrivialResidualFilter`)

save-buffer replied:
Added a test to cover it.

westonpace commented on cpp/src/arrow/compute/exec/hash_join.cc:
```diff
@@ -282,6 +279,125 @@ class HashJoinBasicImpl : public HashJoinImpl {
+        ExecBatch right_key,
+        hash_table_keys_.Decode(match_right.size(), match_right.data()));
+
+    ExecBatch left_payload;
+    if (!schema_mgr_->LeftPayloadIsEmpty()) {
+      ARROW_ASSIGN_OR_RAISE(left_payload, local_state.exec_batch_payloads.Decode(
+                                              match_left.size(), match_left.data()));
+    }
+
+    ExecBatch right_payload;
+    if (!schema_mgr_->RightPayloadIsEmpty()) {
+      ARROW_ASSIGN_OR_RAISE(right_payload, hash_table_payloads_.Decode(
+                                               match_right.size(), match_right.data()));
+    }
```
Do your tests cover situations where the `left_payload` or `right_payload` is empty?

save-buffer replied:
Yep! I actually only knew to put these `if`s in because the Random test caught it

westonpace reviewed (commented):
Looks pretty good.  I'm surprised this only required one additional test as there is a fair amount of nuance here but it is possible that some of these situations are being covered by the existing tests.

save-buffer commented:
The failures seem to be due to conda having a bad time:
```
UserAgent: aws-sdk-cpp/1.8.74/S3/Linux/5.11.0-1020-azure x86_64 GCC/7.5.0
Error: mkdir /tmp/s3fs-test-4m9eos07/bucket/somefile/: file exists (*fs.PathError)
       4: cmd/api-errors.go:2082:cmd.toAPIErrorCode()
       3: cmd/api-errors.go:2107:cmd.toAPIError()
       2: cmd/object-handlers.go:1801:cmd.objectAPIHandlers.PutObjectHandler()
       1: net/http/server.go:2046:http.HandlerFunc.ServeHTTP()
```

Don't think it's related to me. I'll try rebasing in a couple days and maybe the checks will pass.

westonpace commented on cpp/src/arrow/compute/exec/hash_join.cc:
```diff
@@ -282,6 +279,126 @@ class HashJoinBasicImpl : public HashJoinImpl {
+        return Status::OK();
+      }
+    }
+    ARROW_DCHECK(mask.array()->offset == 0);
```
```suggestion
    ARROW_DCHECK_EQ(mask.array()->offset, 0);
```

westonpace commented on cpp/src/arrow/compute/exec/options.h:
```diff
@@ -244,6 +248,8 @@ class ARROW_EXPORT HashJoinNodeOptions : public ExecNodeOptions {
   std::string output_prefix_for_left;
   // prefix added to names of output fields coming from right input
   std::string output_prefix_for_right;
+  // residual filter
```
```suggestion
  // residual filter which is applied to matching rows.  Rows that do not match
  // the filter are not included.  The filter is applied against the
  // concatenated input schema (left fields then right fields) and can reference
  // fields that are not included in the output.
```

westonpace reviewed (approved):
Looks good, just a few nits from me.

westonpace commented:
@bkietz did you want to look over this one more time?  Otherwise I think we can merge.

lidavidm commented on cpp/src/arrow/compute/exec/hash_join_node_test.cc:
```diff
@@ -382,19 +382,24 @@ struct RandomDataTypeVector {
+    return ss.str();
+  }
+
   void Print() {
```
Sorry - a few things:
- Both DataTypeToString and Print are unused.
- SCOPED_TRACE doesn't replace printing; instead inside the scope it's declared/until it's dropped, any test failure will include the string passed to it. So it doesn't make sense to use standalone here. Instead you'd probably have something like `SCOPED_TRACE(Print())` (assuming Print() returns std::string instead of printing) down below in an appropriate location.

save-buffer replied:
Ah yeah seems that there was a print before that used these but then after rebasing those prints got  deleted. I'll just remove these functions

save-buffer commented:
```
[ RUN      ] TestArrowReadDeltaEncoding.DeltaByteArray
unknown file: Failure
C++ exception with description "IOError: Failed to open local file '/arrow/cpp/submodules/parquet-testing/data/delta_byte_array.parquet'. Detail: [errno 2] No such file or directory" thrown in the test body.
```
These failures don't appear to be me.

lidavidm commented:
> ```
> [ RUN      ] TestArrowReadDeltaEncoding.DeltaByteArray
> unknown file: Failure
> C++ exception with description "IOError: Failed to open local file '/arrow/cpp/submodules/parquet-testing/data/delta_byte_array.parquet'. Detail: [errno 2] No such file or directory" thrown in the test body.
> ```
> 
> These failures don't appear to be me.

Hmm, it seems this PR modifies the parquet-testing submodule - perhaps it needs to be rebased/something happened during the last rebase?

save-buffer commented:
Yeah that seems like what happened. I guess I still need some practice with rebasing 😅 I went with the nuclear option and squashed my commits and removed the change to the testing directory, so hopefully it'll be good now.

lidavidm commented:
Ok, looks like these last two failures/timeouts are not related. Thanks for the updates!
